ls1intum/Artemis
# [Feature] Option to recreate build plans on import

### Checklist

### Motivation and Context


Artemis evolves rapidly thanks to numerous contributors. On the flip side, instructors want to reuse older programming exercises e.g. from last semester and import them into their latest courses. This PR enables instructors to recreate and not copy the build plans on import so that they can benefit from the latest Artemis features. Another PR will tackle the problem of intelligently updating the repository so that instructors can benefit from the latest updates to the project structure and dependencies.

### Description


- Refactoring in `ProgrammingExerciseService.createExercise`: Add methods to generate build plan ids and repository names in one place. Reduce complexity of method signatures.
- Add RequestParameter recreateBuildPlans to ProgrammingExerciseResource.importExercise. If true, new build plans are generated and not copied from the imported exercise.
- Add _Recreate Build Plans_ checkbox to the import view.
- Add tests for the new option. Improve the mocking structure a bit.

### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration
3. Create a programming exercise (exercise A)
--> Check that everything works. Also participate and submit
4. Change Template and Solution build plan e.g. add a task by copying the default _Maven Test_ task
5. Import the exercise A with the recreate build plans option activated
--> Check that everything works. Also participate and submit
--> Check the build plans. The build plans should not contain the additional tasks added in step 4
6. Import the exercise A without the recreate build plans option activated
--> Check the build plans. The build plans should contain the additional tasks added in step 4

### Test Coverage

New import logic is fully covered for each programming language

### Screenshots


![image](https://user-images.githubusercontent.com/16407766/98029706-3c01e800-1e10-11eb-8d1b-1b034d5805e2.png)
